src/gluonnlp/utils/parameter.py

```diff
@@ -16,16 +16,90 @@
 # under the License.
 """Utility functions for trainer and parameters."""
 
-__all__ = ['clip_grad_global_norm', 'save_parameters',
+__all__ = ['grad_global_norm', 'clip_grad_global_norm', 'save_parameters',
            'save_states', 'load_parameters', 'load_states']
 
 import warnings
 
-import numpy as np
+from collections import defaultdict
+import mxnet as mx
 from mxnet import nd
 from .. import _constants as C
 from .files import _TempFilePath, _transfer_file_s3
 
+def grad_global_norm(parameters, max_norm=None):
+    """Calculate the 2-norm of gradients of parameters, and how much they should be scaled down
+    such that their 2-norm does not exceed `max_norm`, if `max_norm` if provided.
+
+    If gradients exist for more than one context for a parameter, user needs to explicitly call
+    ``trainer.allreduce_grads`` so that the gradients are summed first before calculating
+    the 2-norm.
+
+    .. note::
+        This function is only for use when `update_on_kvstore` is set to False in trainer.
+
+    Example::
+
+        trainer = Trainer(net.collect_params(), update_on_kvstore=False, ...)
+        for x, y in mx.gluon.utils.split_and_load(X, [mx.gpu(0), mx.gpu(1)]):
+            with mx.autograd.record():
+                y = net(x)
+                loss = loss_fn(y, label)
+            loss.backward()
+        trainer.allreduce_grads()
+        norm = grad_global_norm(net.collect_params().values())
+        ...
+
+    Parameters
+    ----------
+    parameters : list of Parameters
+    max_norm: NDArray, optional
+        The maximum L2 norm threshold. If provided, `ratio` and `is_finite` will be returned.
+
+    Returns
+    -------
+    NDArray
+      Total norm. Shape is (1,)
+    NDArray
+      Ratio for rescaling gradients based on max_norm s.t. grad = grad / ratio.
+      If total norm is NaN, ratio will be NaN, too.
+      Returned if `max_norm` is provided. Shape is (1,)
+    NDArray
+      Whether the total norm is finite, returned if `max_norm` is provided. Shape is (1,)
+    """
+    # distribute gradients among contexts
+    idx = 0
+    arrays = defaultdict(list)
+    sum_norms = []
+    for p in parameters:
+        if p.grad_req != 'null':
+            p_grads = p.list_grad()
+            arrays[idx % len(p_grads)].append(p_grads[idx % len(p_grads)])
+            idx += 1
+    assert len(arrays) > 0, 'No parameter found available for gradient norm.'
+
+    ctx, dtype = arrays[0][0].context, 'float32'
+    for idx, arr in enumerate(arrays.values()):
+        sum_norm = mx.nd.multi_sum_sq(*arr, num_arrays=len(arr))
+        sum_norm = nd.add_n(*sum_norm)
+        sum_norms.append(sum_norm.as_in_context(ctx))
+
+    # reduce
+    total_norm = nd.add_n(*sum_norms).sqrt()
+    if not max_norm:
```

Comment:
can we change this to `if max_norm is None`? The difference is that, `if not max_norm` will evaluate the statement, implicitly invoking asscalar() on `max_norm` which blocks the python main thread.